src/Extensions/SwiftTypeCrawlerExtension.php

```diff
@@ -0,0 +1,313 @@
+
+            // If this record is not published, or we're unable to get a "Live Link" (for whatever reason), then there
+            // is nothing more we can do here
+            if (!$link) {
+                return;
+            }
+
+            // Nothing for us to do here, the Link is already being tracked
+            if (in_array($link, $urls[$key])) {
+                return;
+            }
+
+            // Add our base URL to this key
+            $urls[$key][] = $link;
+        });
+
+        // Update the Urls we have stored for indexing
+        $this->setUrlsToCrawl($urls);
+    }
+
+    /**
+     * Send requests to Swiftype to reindex each of the Urls that we have previously collated
+     */
+    protected function processCollatedUrls(): void
+    {
+        // Fetch the Urls that we need to reindex
+        $key = $this->getOwnerKey();
+
+        // We can't do anything if we don't have a key to process
+        if ($key === null) {
+            return;
+        }
+
+        $urls = $this->getUrlsToCrawl();
+
+        // There is nothing for us to do here if there are no Urls
+        if (count(array_keys($urls)) === 0) {
+            return;
+        }
+
+        // There are no Urls for this particular key
+        if (!array_key_exists($key, $urls)) {
+            return;
+        }
+
+        // Force the reindexing of each URL we collated
+        foreach ($urls[$key] as $url) {
+            $this->forceSwiftypeIndex($url);
+        }
+    }
+
+    protected function forceSwiftypeIndex(string $updateUrl): bool
+    {
+        // We don't reindex dev environments
+        if (Director::isDev()) {
+            return true;
+        }
+
+        $crawler = SwiftypeCrawler::create();
+
+        return $crawler->send($updateUrl);
+    }
+
+    protected function getOwnerKey(): ?string
+    {
+        $owner = $this->getOwner();
+
+        // Can't generate a key if the owner has not yet been written to the DB
+        if (!$owner->isInDB()) {
+            return null;
+        }
+
+        $key = str_replace('\\', '', $owner->ClassName . $owner->ID);
+
+        return $key;
+    }
+
+    /**
+     * Return the link to be indexed, by default this will call {@link AbsoluteLink()} on the published/live
+     * record if it exists. Can be overriden to use a different linking method.
+     *
+     * @throws \InvalidArgumentException if owner does not implement {@link AbsoluteLink()}
+     */
+    protected function getOwnerLink(): ?string
+    {
+        if (!method_exists($this->getOwner(), 'AbsoluteLink')) {
+            throw new \InvalidArgumentException('Invalid object passed, object must implement AbsoluteLink()');
+        }
+
+        $live = Versioned::get_by_stage(get_class($this->getOwner()), Versioned::LIVE)->byID($this->getOwner()->ID);
+
+        if ($live) {
+            $link = $live->AbsoluteLink();
+        } else {
+            $link = null;
+        }
+
+        return $link;
+    }
```

Comment:
I think an `abstract` method would be more appropriate here. The idea of this class should now be that you really have no idea what sort of Model this class will be extended to serve, so any code that isn't "one size fits all" should belong to the classes that extend this `abstract`.

This code can then be moved into `SwiftypeFileCrawlerExtension`, and you've already covered the `SwiftypeSiteTreeCrawlerExtension`.